home/modules/alerts/pages/cve-2022-30331.adoc

```diff
@@ -40,5 +41,22 @@ Only users trusted to review and install custom code should be given this privil
 * Superusers should treat UDFs as developed software and we recommend utilizing the Secure Software Development Life Cycle processes, including static code analysis and appropriate code reviews, prior to installing them on a TigerGraph deployment.
 * If GSQL is configured to xref:gsql-ref:querying:func/query-user-defined-functions.adoc#_use_github_to_store_udfs[read from a GitHub repository for UDFs], ensure that the repository is properly secured, and the hosted code and any future changes are reviewed by system administrators prior to installing them on a TigerGraph deployment.
 
+=== Remediations
+
+==== Blocking UDF Uploading (v3.8.0+)
+
+There are two possible ways for a user to upload a UDF. Starting with TigerGraph 3.8.0, both of those methods are disabled by default, each of them can only be enabled by a user with administrator privilege using the `gadmin` command. For more details, see xref:gsql-ref:querying:func:query-user-defined-functions.adoc#_upload_udfs [Upload UDFs].
+
+==== UDF File Scanning (v3.9.0+)
+
+TigerGraph 3.9.0 introduced a set of checks and restrictions, collectively known as UDF File Scanning, which block certain C{pp} code from being loaded. If this feature is enabled (the default case), then the code of all UDF source code is scanned. UDFs which do not satisfy all the restrictions are blocked from being compiled and run.  For full details, see TigerGraph documentation for xref:tigergraph-server:security.index.adoc#_udf_file_scanning [UDF file scanning].
+
+* *Feature blocklist*. C{pp} headers and features in this private, non-modifiable list are blocked from use.
+
+* *File header allowlist*. Only headers present in the allowlist can be included in UDF files. The default list (see xref:tigergraph-server:security.index.adoc#_udf_file_scanning [UDF file scanning]) can be modified with a `gadmin config set` command.
```

Comment:
may be re-emphasize administrator privileges here.
"can only be modified with a `gadmin config set` command by the system administrator"

Reply:
Thank you. Now I know you are also a good writer.

Reply:
haha. I think it is contagious..